source/iterator/snapshot.go

```diff
@@ -0,0 +1,176 @@
+
+// SnapshotParams is an incoming params for the [NewSnapshot] function.
+type SnapshotParams struct {
+	HubSpotClient *hubspot.Client
+	Resource      string
+	BufferSize    int
+	PollingPeriod time.Duration
+	Position      *Position
+}
+
+// NewSnapshot creates a new instance of the [Snapshot].
+func NewSnapshot(ctx context.Context, params SnapshotParams) *Snapshot {
+	snapshot := &Snapshot{
+		hubspotClient: params.HubSpotClient,
+		resource:      params.Resource,
+		bufferSize:    params.BufferSize,
+		pollingPeriod: params.PollingPeriod,
+		records:       make(chan sdk.Record, params.BufferSize),
+		errC:          make(chan error, 1),
+		position:      params.Position,
+	}
+
+	go snapshot.poll(ctx)
+
+	return snapshot
+}
+
+// HasNext returns a bool indicating whether the iterator has the next record to return or not.
+func (s *Snapshot) HasNext(ctx context.Context) (bool, error) {
+	return len(s.records) > 0, nil
+}
+
+// Next returns the next record.
+func (s *Snapshot) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
+
+	case err := <-s.errC:
+		return sdk.Record{}, fmt.Errorf("async error: %w", err)
+
+	case record := <-s.records:
+		return record, nil
+	}
+}
+
+// poll polls items at the specified time intervals.
+func (s *Snapshot) poll(ctx context.Context) {
+	ticker := time.NewTicker(s.pollingPeriod)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-ticker.C:
+			if err := s.loadRecords(ctx); err != nil {
+				s.errC <- fmt.Errorf("load records: %w", err)
+			}
+		}
+	}
+}
+
+// loadRecords retrieves a new list of the iterator's resource items.
+func (s *Snapshot) loadRecords(ctx context.Context) error {
+	listOpts := &hubspot.ListOptions{
+		Limit: s.bufferSize,
+	}
+
+	if s.position != nil {
+		// add one here in order to skip
+		// this particular element and start from the next one.
+		listOpts.After = strconv.Itoa(s.position.LastID + 1)
+	}
+
+	listResponse, err := s.hubspotClient.List(ctx, s.resource, listOpts)
+	if err != nil {
+		return fmt.Errorf("list %q items: %w", s.resource, err)
+	}
+
+	for _, item := range listResponse.Results {
+		s.position, err = s.getItemPosition(item)
+		if err != nil {
+			return fmt.Errorf("get item's position: %w", err)
+		}
+
+		sdkPosition, err := s.position.MarshalSDKPosition()
+		if err != nil {
+			return fmt.Errorf("marshal sdk position: %w", err)
+		}
+
+		metadata, err := s.getItemMetadata(item)
+		if err != nil {
+			return fmt.Errorf("get item's metadata: %w", err)
+		}
+
+		s.records <- sdk.Util.Source.NewRecordSnapshot(
+			sdkPosition, metadata, sdk.StructuredData{"id": s.position.LastID}, sdk.StructuredData(item),
+		)
+	}
+
+	return nil
+}
+
+// getItemPosition grabs an id field from a provided item and constructs a [Position] based on its value.
+func (s *Snapshot) getItemPosition(item map[string]any) (*Position, error) {
+	itemIDStr, ok := item["id"].(string)
```

Comment:
could you please create constant for "id" variable

Reply:
I've created a constant for the `id` field.